crates/pipeline_manager/src/db/mod.rs

```diff
@@ -1283,174 +1287,89 @@ impl Storage for ProjectDB {
 
-        // Check if we actually changed something before writing a new revision
-        //
-        // Note: What fields are checked is ultimately determined by whatever is
-        // used by the pipeline configuration, e.g., what the `start` function
-        // uses in `runner.rs` to write the config/metadata:
-        if prev_revision.is_some() {
-            let change = txn
-                .query_one(
-                    &compute_revision_current_diff,
-                    &[&pipeline_id.0, &prev_revision],
-                )
-                .await?;
-
-            let nothing_changed = change.get::<_, i64>(0) == 0;
-            if nothing_changed {
+        if let Some(row) = row {
+            let current_revision: String = row.get(0);
+            if pipeline_config == current_revision {
                 return Err(DBError::RevisionNotChanged);
             }
         }
-
-        // TODO: Ideally these queries happen in the same transaction `txn`.
-        // Need to revise the API a bit to optionally pass in a transaction
-        // object for this to work seamlessly. Also probably the program row
-        // should be locked with FOR UPDATE while reading to prevent the
-        // compiler from resetting the status field.
-        let (_pipeline, program, _connectors) =
-            self.pipeline_is_committable(tenant_id, pipeline_id).await?;
-
-        // Copy all pipeline data to history tables
-        //
-        // TODO(performance): In theory the following inserts could all run in
-        // parallel with async but I couldn't figure out how to make it work
-        // with the args :/
-        txn.execute(&insert_program_history, &[&revision, &program.program_id.0])
-            .await?;
-        txn.execute(&insert_pipeline_history, &[&revision, &pipeline_id.0])
-            .await?;
-        txn.execute(&insert_connector_history, &[&revision, &pipeline_id.0])
+        let stmt = txn
+            .prepare_cached("INSERT INTO deployed_pipeline (id, pipeline_id, tenant_id, config) VALUES ($1, $2, $3, $4)
+                             ON CONFLICT ON CONSTRAINT unique_pipeline_id 
+                             DO UPDATE SET id = EXCLUDED.id, config = EXCLUDED.config")
             .await?;
         txn.execute(
-            &insert_attached_connector_history,
-            &[&revision, &pipeline_id.0],
+            &stmt,
+            &[
+                &deployment_id,
+                &pipeline_id.0,
+                &tenant_id.0,
+                &pipeline_config,
+            ],
         )
         .await?;
-
-        // Update the revision of the pipeline object
-        txn.execute(&update_revision, &[&revision, &pipeline_id.0, &tenant_id.0])
-            .await?;
-
         txn.commit().await?;
-
-        Ok(Revision(revision))
+        Ok(Revision(deployment_id))
     }
 
-    async fn get_last_committed_pipeline_revision(
+    async fn get_current_pipeline_revision(
```

Comment:
get_latest_pipeline_deployment (note latest instead of last/current)

Reply:
I'll keep it simple with get_pipeline_deployment (since there can only be 0 or 1 of them).